### PR TITLE
feat(arch/riscv): add SBI hart suspend

### PR DESCRIPTION
**Pull request by joaopeixoto13**

### CODE REVIEW

File: src/arch/riscv/sbi.c [hunk 1]
```diff
@@ -382,6 +389,35 @@ static struct sbiret sbi_hsm_status_handler(void)
...
+    } else {
+        if (suspend_type & SBI_HSM_SUSP_NON_RET_BIT) {
+            // spin_lock(&cpu()->vcpu->arch.sbi_ctx.lock);
+            // cpu()->vcpu->arch.sbi_ctx.start_addr = resume_addr;
+            // cpu()->vcpu->arch.sbi_ctx.priv = (unsigned int)priv;
+            // spin_unlock(&cpu()->vcpu->arch.sbi_ctx.lock);
+
+            /*
+             * For now we're not supporting non-retentive suspend since we
+             * will need to save and later restore various hart registers and CSRs.
```
Comment (josecm) [inline src/arch/riscv/sbi.c:411]: Can we add a TODO here saying that we only need to implement this when we get a real physical platform with  real non-retentive suspend implemented?

**Review by josecm — CHANGES_REQUESTED** [review]
SBI HSM suspend operation is only supported from SBI 0.3 onward, so you should also update the version expected from the firmware and the one return to the guest.

File: src/arch/riscv/sbi.c [hunk 2]
```diff
@@ -382,6 +389,35 @@ static struct sbiret sbi_hsm_status_handler(void)
...
+
+    spin_lock(&cpu()->vcpu->arch.sbi_ctx.lock);
+    if (cpu()->vcpu->arch.sbi_ctx.state != STARTED) {
+        ret.error = SBI_ERR_FAILURE;
+    } else {
+        if (suspend_type & SBI_HSM_SUSP_NON_RET_BIT) {
+            // spin_lock(&cpu()->vcpu->arch.sbi_ctx.lock);
+            // cpu()->vcpu->arch.sbi_ctx.start_addr = resume_addr;
+            // cpu()->vcpu->arch.sbi_ctx.priv = (unsigned int)priv;
+            // spin_unlock(&cpu()->vcpu->arch.sbi_ctx.lock);
```
[the remark this answers is not in the document]
Comment (joaopeixoto13) [reply src/arch/riscv/sbi.c]: > Please don't keep commented out code in the commit.

Done [here](https://github.com/bao-project/bao-hypervisor/compare/5f663e3d4afbf5c9ff81ced7fd05a6a259bba205..170912648fb06cf85cf9853b409920926d3b3987)

File: src/arch/riscv/sbi.c [hunk 3]
```diff
@@ -382,6 +389,35 @@ static struct sbiret sbi_hsm_status_handler(void)
...
+
+            /*
+             * For now we're not supporting non-retentive suspend since we
+             * will need to save and later restore various hart registers and CSRs.
+             */
+            ERROR("non-retentive suspend not supported");
```
[the remark this answers is not in the document]
Comment (joaopeixoto13) [reply src/arch/riscv/sbi.c]: > According to the SBI spec you should just return `SBI_ERR_NOT_SUPPORTED`

Done [here](https://github.com/bao-project/bao-hypervisor/compare/5f663e3d4afbf5c9ff81ced7fd05a6a259bba205..170912648fb06cf85cf9853b409920926d3b3987)

File: src/arch/riscv/sbi.c (hunk 1 above)
Comment (joaopeixoto13) [reply src/arch/riscv/sbi.c]: > Can we add a TODO here saying that we only need to implement this when we get a real physical platform with real non-retentive suspend implemented?

Done [here](https://github.com/bao-project/bao-hypervisor/compare/5f663e3d4afbf5c9ff81ced7fd05a6a259bba205..170912648fb06cf85cf9853b409920926d3b3987)

**Review by josecm — APPROVED** [review]

**Pull request merged**